File: src_qt/rules/ConditionVariableNode.cpp
#include <QHashIterator>

#include "ConditionVariableNode.h"


namespace CnotiMind
{

	ConditionVariableNode::ConditionVariableNode( const QString& variable, const QString& value, ConditionOperator op, Brain* brain, QObject* parent ) :
		ConditionNode( op, brain, parent ),
		_variable( variable ),
		_value( value ),
		_valueNumeric( value.toInt( &_isValueNumeric ) )
	{
	}

	void ConditionVariableNode::exec()
	{
		// Does nothing, since this Node, need variables to exec
	}

	void ConditionVariableNode::exec( QHash<QString,QString>& variables )
	{
		if( isTrue( variables ) )
		{
			execChildren( variables );
		}
	}

	/*
		Returns false, since it requires variables to work correctly
	*/
	bool ConditionVariableNode::isTrue()
	{
		return false;
	}

	QString ConditionVariableNode::info( int depth ) const
	{
		QString info;

		info += space(depth) + "Condition type=Variable";

		return info + RuleNode::info(depth);
	}

	/*
		Test the value of the variable using the conditional operator
	*/
	bool ConditionVariableNode::isTrue( QHash<QString,QString>& variables )
	{
		QHashIterator<QString, QString> it( variables );
		while( it.hasNext() )
		{
			it.next();
			const QString& k = it.key();
			const QString& v = it.value();

			// Test if found the variable
			if( _variable.compare( k, Qt::CaseInsensitive ) == 0 )
			{
				// Try to convert the value from the variable to number
				bool ok;
				qreal value = v.toDouble(&ok);

				if( ok ) // if the variable value is a number
				{
					if( _isValueNumeric ) // if the value to compare is a number
					{
						switch( _operator )
						{
							case ConditionOperatorBigger: return value > _valueNumeric ;
							case ConditionOperatorBiggerOrEqual: return value >= _valueNumeric;
							case ConditionOperatorSmaller: return value < _valueNumeric;
							case ConditionOperatorSmallerOrEqual: return value <= _valueNumeric;
							case ConditionOperatorEqual: return value == _valueNumeric;
							case ConditionOperatorDifferent: return value != _valueNumeric;
						}
						return false;
					}
					else // if it is not a number, test if it is variable
					{
						QString strValue = _value;
						tagsToValue( strValue, variables ); // convert variable tags to number

						qreal compareValue = strValue.toDouble(&ok);

						if(ok) // if it is number
						{
							switch( _operator )
							{
								case ConditionOperatorBigger: return value > compareValue;
								case ConditionOperatorBiggerOrEqual: return value >= compareValue;
								case ConditionOperatorSmaller: return value < compareValue;
								case ConditionOperatorSmallerOrEqual: return value <= compareValue;
								case ConditionOperatorEqual: return value == compareValue;
								case ConditionOperatorDifferent: return value != compareValue;
							}
							return false;
						}
						// if it is not a number use the code below
					}
				}

				// If one of the value is not a number
				QString strValue = _value;
				tagsToValue( strValue, variables );

				switch( _operator ) // it's a string, just to this 2 operators
				{
					case ConditionOperatorEqual: return QString::compare( strValue, v, Qt::CaseInsensitive ) == 0;
					case ConditionOperatorDifferent: return QString::compare( strValue, v, Qt::CaseInsensitive ) != 0;
				}
				return false;
			}
		}

		// No equal variables found
		return false;
	}

	ConditionVariableNode* ConditionVariableNode::fromXML(const QString &qName, const QXmlAttributes &atts, Brain *brain, QObject *parent)
	{
		if(qName.compare( "Condition", Qt::CaseInsensitive) == 0)
		{
			QString type = atts.value( "type" );
			if( type.compare("Variable", Qt::CaseInsensitive ) == 0 )
			{
				QString key = atts.value( "variable" );
				QString value = atts.value( "compareValue" );
				ConditionOperator op = translateConditionOperator( atts.value( "operator" ) );

				return new ConditionVariableNode( key, value, op, brain, parent );

			}
		}
		return NULL;
	}

}

File: src_ios/CnotiMind/rules/ConditionVariableNode.m
//
//  ConditionEmotionNode.m
//  CnotiMind
//
//  Created by Gonçalo Rodrigues on 11/05/09.
//  Copyright 2011 Cnotinfor. All rights reserved.
//

#import "ConditionVariableNode.h"

#import "Brain.h"
#import "Perception.h"

@implementation ConditionVariableNode


- (id) initWithVariableAndValueAndOperatorAndBrainAndParent: (NSString*)aKey 
                                                      value:(NSString*)aValue 
                                                   operator: (enum ConditionOperator)aOperator 
                                                      brain:(Brain*)aBrain 
                                                     parent:(id)aParent
{
    
    if (self == [super initWithKeyAndValueAndBrainAndParent: aKey 
                                                      value:aValue 
                                                   operator:aOperator 
                                                      brain:aBrain 
                                                     parent:aParent])
    {
    }
    return self;
}


- (void) exec
{
    // Does nothing, since this Node, need variables to exec
}


- (void) exec:(NSMutableDictionary*)aVariables
{
    if ([self isTrue:aVariables]) {
        [aVariables setObject:_value forKey:@"[perception.value]"];
        [self execChildren:aVariables];
    }

}


- (NSString*) info:(int)aDepth
{
    NSString* info = [NSString stringWithFormat:@""];
    NSString* space = [self space:aDepth];
    
    info = [info stringByAppendingFormat:@"%@ Condition type=Variable", space];
    info = [info stringByAppendingFormat:@"%@", [super info:aDepth]];
    
    return info;
}

/*
 Returns false, since it requires variables to work correctly
 */
- (BOOL) isTrue
{
    return FALSE;
}


- (BOOL) isTrue:(NSMutableDictionary*)aVariables
{
    NSEnumerator* eVariables = [aVariables keyEnumerator];
    
    for (NSString* aKey in eVariables) {
        
        NSString* k = [NSString stringWithFormat:@"%@", aKey];
        NSString* v = [NSString stringWithFormat:@"%@", [aVariables objectForKey:aKey]];

        // Test if found the variable
        if (![_key caseInsensitiveCompare:k]) {
            
            // Try to convert the value from the variable to number
            float value = [v floatValue];
//            BOOL ok = ([[NSScanner scannerWithString:v] scanFloat:NULL]) ? TRUE : FALSE;
            
            // if both values are numbers
            if ([self isNumeric:v]) {
                if (_isValueNumeric) { // if the value to compare is a number
                    
                    switch( (int)_operator )
                    {
                        case ConditionOperatorBigger: return  value > _valueNumeric;
                        case ConditionOperatorBiggerOrEqual: return value >= _valueNumeric;
                        case ConditionOperatorSmaller: return value < _valueNumeric;
                        case ConditionOperatorSmallerOrEqual: return value <= _valueNumeric;
                        case ConditionOperatorEqual: return value == _valueNumeric;
                        case ConditionOperatorDifferent: return value != _valueNumeric;
                    }
                    return false;    
                }
                
                else { // if it is not a number, test if it is variable
                
                    NSString* strValue = [NSString stringWithFormat:@"%@", _value];
                    [self tagsToValue:&strValue variables:&aVariables];
                    
                    if ([self isNumeric:strValue]) {
                        float compareValue = [strValue floatValue];

                        switch( (int)_operator )
                        {
                            case ConditionOperatorBigger: return value > compareValue;
                            case ConditionOperatorBiggerOrEqual: return value >= compareValue;
                            case ConditionOperatorSmaller: return value < compareValue;
                            case ConditionOperatorSmallerOrEqual: return value <= compareValue;
                            case ConditionOperatorEqual: return value == compareValue;
                            case ConditionOperatorDifferent: return value != compareValue;
                        }
                        return false;
                    }
                }           
            }
        
            else // If one of the value is not a number
            {
                NSString* strValue = [NSString stringWithFormat:@"%@", _value];
                [self tagsToValue:&strValue variables:&aVariables];                
                
                switch( (int)_operator ) // it's a string, just to this 2 operators
                {
                    case ConditionOperatorEqual: return ![strValue caseInsensitiveCompare:v];
                    case ConditionOperatorDifferent: return [strValue caseInsensitiveCompare:v];
                }
                return false;
            }
        }
    }
    
    // No equal variables found
    return false;
}

@end